ir/src/lib.rs

```diff
@@ -81,47 +88,63 @@ impl AirIR {
                     if trace_bindings.len() > 1 {
                         validator.exists("aux_trace_columns");
                     }
-                    // process & validate the trace bindings
-                    symbol_table.insert_trace_bindings(trace_bindings)?;
+                    // accumulate and save the trace bindings for later processing.
+                    trace_decls.push(trace_bindings);
                 }
                 ast::SourceSection::PublicInputs(inputs) => {
-                    // process & validate the public inputs
-                    symbol_table.insert_public_inputs(inputs)?;
                     validator.exists("public_inputs");
+                    // accumulate and save the public input bindings for later processing.
+                    pub_input_decls.extend(inputs);
                 }
                 ast::SourceSection::PeriodicColumns(columns) => {
                     // process & validate the periodic columns
                     symbol_table.insert_periodic_columns(columns)?;
                 }
                 ast::SourceSection::RandomValues(values) => {
-                    symbol_table.insert_random_values(values)?;
                     validator.exists("random_values");
+                    // save the random value declaration for later processing.
+                    rand_decls.push(values);
```

Comment:
Why not process random values here? (similar to what we do with periodic columns and constants)